flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java

```diff
@@ -630,10 +630,24 @@ JobManagerRunner createJobManagerRunner(JobGraph jobGraph, long initializationTi
 
     @Override
     public CompletableFuture<Acknowledge> cancelJob(JobID jobId, Time timeout) {
-        Optional<JobManagerRunner> maybeJob = getJobManagerRunner(jobId);
+        boolean isGloballyTerminated;
+        try {
+            isGloballyTerminated = jobResultStore.hasJobResultEntry(jobId);
```

Comment:
This makes me thing that the JRS interface should have been asynchronous as we can potentially block the RPC threads here 🤔

Reply:
Other thought would be, that unlike the previous solution, this safeguard only works with globally terminal states